### ISSUE TITLE
tree_util register_X functions remove type hints on decorated classes

### ISSUE THREAD

**Issue by bionicles** [google/jax#22052]
### Description

Thank you for working on JAX!

TLDR: tree_util docs recommend use of decorators; however, **tree_util register_X functions lack return type hints, so any types you decorate become typed NoneType**. Not sure of implications of that, but it seems like a bug if we want nice type hints!

Seems like there are 2 fixes, not mutually exclusive, 

A) update docs [1] to show usage examples of calling these as free I/O functions on types (so the type added to the registry isn't defined by the jax function)
B) update tree_util register functions to return the same type as their input

This bug requires Python 3.12 to reproduce, as it relies on newer type system generic features. 

```py
from typing import Protocol
from dataclasses import dataclass
from functools import partial

import jax


class Api(Protocol):
    pass


@partial(
    jax.tree_util.register_dataclass,  # returns None
    meta_fields=["metadata_a"],
    data_fields=["data"],
)
@dataclass(frozen=True, slots=True, kw_only=True)
class A[GENERIC: Api]:
    metadata_a: GENERIC
    data: jax.Array


@dataclass(frozen=True, slots=True, kw_only=True)
class B[GENERIC: Api]:
    metadata_a: GENERIC
    data: jax.Array


jax.tree_util.register_dataclass(
    B,
    meta_fields=["metadata_a"],
    data_fields=["data"],
)


def test_dataclass_registration_decorator_type_hints():
    type_a = type(A)
    type_b = type(B)
    print(f"{type_a=}")
    print(f"{type_b=}")
    assert type_b == type
    assert type_a == type


if __name__ == "__main__":
    jax.print_environment_info()
    test_dataclass_registration_decorator_type_hints()
```
![image](https://github.com/google/jax/assets/24532336/79ce1edb-06b7-4669-a327-c853ce40486a)

### System info (python version, jaxlib version, accelerator, etc.)

```sh
jax:    0.4.28
jaxlib: 0.4.28
numpy:  1.26.4
python: 3.12.0 | packaged by conda-forge | (main, Oct  3 2023, 08:43:22) [GCC 12.3.0]
jax.devices (1 total, 1 local): [cuda(id=0)]
process_count: 1
platform: uname_result(system='Linux', node='WIN-QVRBL09D89C', release='5.15.153.1-microsoft-standard-WSL2', version='#1 SMP Fri Mar 29 23:14:13 UTC 2024', machine='x86_64')


$ nvidia-smi
Sun Jun 23 12:18:58 2024       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 555.52.01              Driver Version: 555.99         CUDA Version: 12.5     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA GeForce RTX 4090        On  |   00000000:01:00.0  On |                  Off |
|  0%   49C    P2             33W /  450W |    1533MiB /  24564MiB |     13%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
                                                                                         
+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI        PID   Type   Process name                              GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|    0   N/A  N/A   1268474      C   /python3.12                                 N/A      |
+-----------------------------------------------------------------------------------------+
```

**Comment by bionicles** [comment]
ah, whoops, should be
assert type_b == type
assert type_a == type
i.e. the type of these classes ought to be type. using the tree_util decorator makes it NoneType, causing generics to fail

(updated)

**Comment by jakevdp** [comment]
It looks like you need to update your JAX version. `register_dataclass` returned `None` before JAX v0.4.29, so you cannot use the decorator approach in older JAX versions. If you want to use the decorator approach from the docs, you'll have to update to JAX v0.4.29 or newer.

If you can't update to a more recent JAX version, then you can register your class this way:
```python
@dataclass(frozen=True, slots=True, kw_only=True)
class A[GENERIC: Api]:
    metadata_a: GENERIC
    data: jax.Array

jax.tree_util.register_dataclass(A, meta_fields=["metadata_a"], data_fields=["data"])
```
